src/opuntia.erl

```diff
@@ -132,27 +146,49 @@ calculate(#token_bucket_shaper{shape = {MaximumTokens, Rate},
     TokensOverused = max(0, TokensNowUsed - ExactlyAvailableNow),
 
     %% And then MaybeDelay will be zero if TokensOverused was zero
-    MaybeDelayMs = TokensOverused / Rate,
+    OverUsedRateNow = TokensOverused / Rate,
 
     %% We penalise rounding up, the most important contract is that rate will never exceed that
     %% requested, but the same way timeouts in Erlang promise not to arrive any time earlier but
     %% don't promise at what time in the future they would arrive, nor we promise any upper bound
     %% to the limits of the shaper delay.
-    RoundedDelayMs = ceil(MaybeDelayMs),
-
-    NewShaper = Shaper#token_bucket_shaper{available_tokens = TokensAvailable,
-                                           last_update = NativeNow + RoundedDelayMs,
-                                           debt = RoundedDelayMs - MaybeDelayMs},
-    {NewShaper, RoundedDelayMs}.
+    case OverUsedRateNow - OverPenalisedInUnitsLastTime of
+        %% Even after paying the old debt you incurr a new debt again
+        Punish when Punish >= +0.0 ->
+            DelayMs = convert_time_unit(Punish, TimeUnit, millisecond),
+            RoundedDelayMs = ceil(DelayMs),
+            OverPenalisedNow = RoundedDelayMs - DelayMs,
+            OverUsedRateNowInUnits = convert_time_unit(OverPenalisedNow, millisecond, TimeUnit),
+            DelayNative = convert_time_unit(RoundedDelayMs, TimeUnit, native),
+            RoundedDelayNative = ceil(DelayNative),
+            NewShaper = Shaper#token_bucket_shaper{available_tokens = TokensAvailable,
+                                                   last_update = NativeNow + RoundedDelayNative,
+                                                   debt = OverUsedRateNowInUnits},
+            {NewShaper, RoundedDelayMs};
+        %% I penalised you too much last time, you get off now but with a future bill
+        Punish when Punish =< +0.0 ->
```

Comment:
This condition has the common value of `+0.0` with the previous one, which is shadowed by the previous clause and looks like a mistake. Also: there is no point in the plus sign, because `-0.0` is equal to `0.0` and `+0.0`.